src/nunit.analyzers/ComparableTypes/ComparableTypesAnalyzer.cs

```diff
@@ -0,0 +1,153 @@
+                if (expectedOperation == null)
+                    continue;
+
+                var expectedType = expectedOperation.Type;
+                if (expectedType == null)
+                    return;
+
+                if (!CanCompare(actualType, expectedType, context.Compilation))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(
+                        descriptor,
+                        expectedOperation.Syntax.GetLocation(),
+                        constraintMethod.Name));
+                }
+            }
+        }
+
+        private static bool CanCompare(ITypeSymbol actualType, ITypeSymbol expectedType, Compilation compilation)
+        {
+            if (actualType.OriginalDefinition.SpecialType == SpecialType.System_Nullable_T)
+                actualType = ((INamedTypeSymbol)actualType).TypeArguments[0];
+
+            if (expectedType.OriginalDefinition.SpecialType == SpecialType.System_Nullable_T)
+                expectedType = ((INamedTypeSymbol)expectedType).TypeArguments[0];
+
+            var conversion = compilation.ClassifyConversion(actualType, expectedType);
+            if (conversion.IsNumeric)
+                return true;
+
+            if (IsIComparable(actualType, expectedType) || IsIComparable(expectedType, actualType))
+                return true;
+
+        /*
+            // NUnit doesn't demand that IComparable is for the same type.
+            if (IsIComparable(actualType) || IsIComparable(expectedType))
+            {
+                // Except for enum
+                return actualType.BaseType.SpecialType != SpecialType.System_Enum || actualType == expectedType;
+            }
+        */
+
+            if (actualType == expectedType && IsIComparable(actualType))
+                return true;
+
+            return false;
+        }
+
+        private static bool IsIComparable(ITypeSymbol typeSymbol, ITypeSymbol comparableTypeArguments)
+        {
+            if (typeSymbol.AllInterfaces.Any(i => i.TypeArguments.Length == 1
+                && i.TypeArguments[0].Equals(comparableTypeArguments)
+                && i.GetFullMetadataName() == "System.IComparable`1"))
+            {
+                return true;
+            }
+
+            // NUnit allows for an CompareTo method, even if not implementing IComparable.
+            return typeSymbol.GetAllMembers().Any(x => x is IMethodSymbol methodSymbol && methodSymbol.Name == "CompareTo"
+                && methodSymbol.Parameters.Length == 1 && methodSymbol.Parameters[0].Type.Equals(comparableTypeArguments));
+        }
+
+        private static bool IsIComparable(ITypeSymbol typeSymbol)
+        {
+            if (typeSymbol.AllInterfaces.Any(i => i.TypeArguments.Length == 0
+                && i.GetFullMetadataName() == "System.IComparable"))
+            {
+                return true;
+            }
+
+            // NUnit allows for an CompareTo method, even if not implementing IComparable.
```

Comment:
Nope, this is not allowed. 
See https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Constraints/NUnitComparer.cs

Reply:
My bad, I extrapolated your example to `CompareTo(object?)`

Reply:
Deleted and allows me to delete a unit tests which tested that.